applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -556,7 +556,39 @@ static void cmd_send(uint8_t *buf, size_t cmd_length, size_t buf_size)
 		rsp_send_error();
 		return;
 	} else if (err > 0) {
-		LOG_ERR("AT command error, type: %d", nrf_modem_at_err_type(err));
+		LOG_ERR("AT command error (%d), type: %d: buf:'%s'", err, nrf_modem_at_err_type(err), buf);
+		err = 0;
+		/* TODO: modem sends these error strings/codes already for AT commands that it's handling.
+		 * This is needed at least for #XSMS where the response is not written into 'buf'
+		 * bute we only have the error code.
+		 *
+		 * If there is some information in 'buf', this will override it. I guess there
+		 * shouldn't be but this is still likely a wrong way to do this.
+		 */
+
+		/* Reconstruct 'ERROR', 'CME ERROR' and 'CMS ERROR' response from
+		 * nrf_modem_at_cmd() return value.
+		 */
+		switch (nrf_modem_at_err_type(err)) {
+		case NRF_MODEM_AT_CME_ERROR:
+			err = snprintf(buf + strlen(CRLF_STR), buf_size - strlen(CRLF_STR),
+				"+CME ERROR: %d\r\n", nrf_modem_at_err(err));
+			break;
+		case NRF_MODEM_AT_CMS_ERROR:
+			err = snprintf(buf + strlen(CRLF_STR), buf_size - strlen(CRLF_STR),
+				"+CMS ERROR: %d\r\n", nrf_modem_at_err(err));
+			break;
+		case NRF_MODEM_AT_ERROR:
+		default:
+			err = snprintf(buf + strlen(CRLF_STR), buf_size - strlen(CRLF_STR),
+				"ERROR\r\n");
+			break;
+		}
+		if (err != 0) {
+			LOG_ERR("Too small buffer for ERROR response %d", err);
+			rsp_send_error();
+			return;
+		}
 	}
```

Comment:
This logic completely discards the final result from modem and constructs it locally.

 * @returns A positive value On "ERROR", "+CME ERROR", and "+CMS ERROR" responses.
 *	    The type of error can be distinguished using @c nrf_modem_at_err_type.
 *	    The error value can be retrieved using @c nrf_modem_at_err.
 
 I'd suggest to adjust _buf_ only when it's not consistent with error type. 
 The #XSMS problem is due to "+CMS ERROR" is lost in libsms where the "+CMGS" modem AT command is issued.

Reply:
Modified version done into `slm_at_cb_wrapper`